crates/flux-fhir-analysis/src/conv/mod.rs

```diff
@@ -1644,10 +1646,60 @@ impl<'genv, 'tcx, P: ConvPhase> ConvCtxt<'genv, 'tcx, P> {
                     .try_collect()?;
                 rty::Expr::adt(def_id, flds)
             }
+            fhir::ExprKind::Constructor(path, exprs, spread) => {
+                let def_id = match path.res {
+                    ExprRes::Struct(def_id) | ExprRes::Enum(def_id) => def_id,
+                    _ => span_bug!(path.span, "unexpected path in constructor"),
+                };
+                let assns = self.conv_constructor_exprs(def_id, env, exprs, spread)?;
+                rty::Expr::adt(def_id, assns)
+            }
         };
         Ok(self.add_coercions(expr, fhir_id))
     }
 
+    fn conv_constructor_exprs(
+        &mut self,
+        struct_def_id: DefId,
+        env: &mut Env,
+        exprs: &[fhir::FieldExpr],
+        spread: &Option<&fhir::Spread>,
+    ) -> QueryResult<List<rty::Expr>> {
+        let struct_adt = self.genv.adt_sort_def_of(struct_def_id)?;
+        let sort_by_field = FxIndexSet::from_iter(struct_adt.field_names());
```

Comment:
We could make `struct_adt.field_names()` return an `&FxIndexSet<Symbol>`. and even more, we can make the field `field_names` in `AdtSortDefData` an `FxIndexSet<Symbol>` instead of a `Vec<Symbol>`.

Also,

```suggestion
        let field_names = FxIndexSet::from_iter(struct_adt.field_names());
```

Reply:
Or maybe we can invert this, map `exprs` by name, and iterate over the field names. Something like

```rust
let spread = spread
    .map(|spread| self.conv_expr(env, &spread.expr))
    .transpose()?;
let field_exprs_by_name: FxIndexMap<Symbol, &fhir::FieldExpr> =
    exprs.iter().map(|e| (e.ident.name, e)).collect();
for (idx, field_name) in struct_adt.field_names().enumerate() {
    if let Some(field_expr) = field_exprs_by_name.get(field_name) {
        assns.push(self.conv_expr(env, &field_expr.expr)?);
    } else if Some(spread) = spread {
        let proj = rty::FieldProj::Adt { def_id: struct_def_id, field: idx as u32 };
        assns.push(rty::Expr::field_proj(spread, proj));
    }
}
```

Reply:
Nice I like that one.